# Numpy Practice

In this task we will work with the following tensors:

1. <img src="images\img.png" width="350" />
2. <img src="images\img_1.png" width="350" />
3. <img src="images\img_2.png" width="350" />
4. <img src="images\img_3.png" width="350" />

In [2]:
import numpy as np
import json_tricks
import lovely_numpy as ln

In [3]:
answer = {}

For the sake of visualization, I will add a vector `x` below on which I will demonstrate how operations wit vectors work. Your task will be to apply necessary fucntions where it is needed to do so.

In [4]:
x = np.array([0, 1, 2, 3, 4, 5]) ** 2

# 1. Build numpy vectors

Firstly let us write numpy for the vectors that are enumerated in the task in form of list of numpy tensors

For instance, for the vector 
$$\mathbf x = 
\begin{bmatrix}
1 \\
2 \\
3 
\end{bmatrix},$$
the answer should be ```np.array([1, 2, 3])```.

In [5]:
answer['vectors'] = [
    np.array([1, 2, 3]),
]


# 2. Get operations with vectors (slicing)

Vectors you can:
- index
    - `x[3]` (returns corresponding coordinate)

- slice, 
    - `x[2:4]` (returns vector of corresponding coordinates from 2nd to 4th (non-inlusively))
    - `x[:3]` (extractsall coordinates up to 3rd)
    - `x[3:]` (extracts coordinates from 3rd an further)

- slice with stride
    - `x[1:5:2]` (extracts all the coordinates from 1st to 5th (non-inclusively) with stride 2)
    - `x[::2]` (extracts every second coordinate)
    - `x[::-1]` (inverts a vector due to -1 stride)

- index a vector with another vector or list
    - `x[[1, 2, 5]]` will output 1st, 2nd and 5th coordinates

- index a vector using a binary mask, for instance
    - `x[x > 4]` will get all of the coordinates that are larger than 4 (`x > 4` returns `[False, False, False, True, True, True]`) 

In [6]:
print('x :\t\t',  x)
print('x[3] :\t\t', x[3])
print('x[2:4] :\t', x[2:4] )
print('x[:3] :\t\t', x[:3])
print('x[3:] :\t\t', x[3:])
print('x[1:5:2] :\t', x[1:5:2])
print('x[2:] :\t\t', x[2:])
print('x[::2] :\t', x[::2])
print('x[::-1] :\t', x[::-1])
print('x[[1, 2, 5]] :\t', x[[1, 2, 5]])
print('x[x > 4] :\t', x[x > 4])


x :		 [ 0  1  4  9 16 25]
x[3] :		 9
x[2:4] :	 [4 9]
x[:3] :		 [0 1 4]
x[3:] :		 [ 9 16 25]
x[1:5:2] :	 [1 9]
x[2:] :		 [ 4  9 16 25]
x[::2] :	 [ 0  4 16]
x[::-1] :	 [25 16  9  4  1  0]
x[[1, 2, 5]] :	 [ 1  4 25]
x[x > 4] :	 [ 9 16 25]


## Task

For exercise, create a list containing:
- 0th coordinate of last vector in the exercise
- last coordinate of last vector in the exercise
- all coordinates with even indices of the last vector in the exercise
- all coordinates with odd indices of the last vector in the exercise
- inverted last vector
- 1st, 0th and 4th coordinates of vector in vector

In [7]:
answer['slices'] = [
    x[0],
    x[-1],
    x[0::2],
    x[1::2],
    x[::-1],
    x[[1,0,4]]
]

# 3. Get operations for vectors

- set and get values with index operators 
    - `x[3] = 1` -- your vector 3rd coordinate will be set to 1
    - `x[2:5] = 1` -- sets all of the coordinates from 2nd to 5th to 1
    - you can use any other indexing method to set the coordinates to a specified value


It is very important to remember that by default numpy copies are **shallow**, meaning that code
```
y = x
y[1] = 3
```

whill change 1st coordinate for vector named `x` as `x` and `y` ar basically the same vector with two names

To avoid this, there is a `x.copy()` method

In [8]:
print('Shallow copy')
print('x', x)
y = x
print('y', y)
y[0] = 10
print('x', x)
print('y', y)

x[0] = 0

print('Deep copy')
y = x.copy()
y[0] = 10
print('x', x)
print('y', y)


Shallow copy
x [ 0  1  4  9 16 25]
y [ 0  1  4  9 16 25]
x [10  1  4  9 16 25]
y [10  1  4  9 16 25]
Deep copy
x [ 0  1  4  9 16 25]
y [10  1  4  9 16 25]


## Task 

For the last vector create deep copies with:
- 3rd coordinate set to 1
- 2-4th coordinates including the last set to 10
- Every even coordinate set to 2
- Every odd coordinate set to 3
- Last coordinate set to -1
- 0, 1 and last coordinates set to -10
- All coordinates that are smaller than 1 should be set to 0

In [29]:

answer['deep_copy'] = [
    x[3]=1,
    x[2:5]=10,
    x[-1]=10,
    x[0::2]=2,
    x[1::2]=3,
    x[-1]=-1
    x[0:2],
    x[x<1]=0
]

SyntaxError: cannot assign to subscript here. Maybe you meant '==' instead of '='? (2919509614.py, line 2)

# 4. Checking properties for Numpy arrays

You can check properties of an array `x` with the following commands:
- `x.shape` checks shape of array
- `x.ndim` checks number of dimensions of an array (is it 1d table or 2d table, or 3d table)
- `x.size` checks number of elements
- `x.dtype` checks data type of the elements
- `x.itemsize` checks number of bits for every element
- `x.nbytes` checks whole memory occupied by the array

In [10]:
print('x.shape : \t', x.shape)
print('x.ndim : \t', x.ndim)
print('x.size : \t', x.size)
print('x.dtype : \t', x.dtype)
print('x.itemsize : \t', x.itemsize)
print('x.nbytes : \t', x.nbytes)


x.shape : 	 (6,)
x.ndim : 	 1
x.size : 	 6
x.dtype : 	 int64
x.itemsize : 	 8
x.nbytes : 	 48


## Task

Check sizes of all arrays in `vectors` and return them as answer

In [11]:
answer['shapes'] = [
    ## YOUR CODE HERE
]

# 5. Shaping arrays

You can flatten or reshape numpy arrays as soon as it has enough elements to fit all the shape

For example, vector `x` can be converted to shape `[2, 3]` and `[3, 2]`

It is possible to replace one of the dimensionalities with `-1` so that it is dynamically determined from the number of elements of initial tensor

In [12]:
print('x.reshape([2, 3]) :\n', x.reshape([2, 3]))
print('x.reshape([3, 2]) :\n', x.reshape([3, 2]))

print('x.reshape([2, -1]) :\n', x.reshape([2, -1]))
print('x.reshape([3, -1]) :\n', x.reshape([3, -1]))

x.reshape([2, 3]) :
 [[ 0  1  4]
 [ 9 16 25]]
x.reshape([3, 2]) :
 [[ 0  1]
 [ 4  9]
 [16 25]]
x.reshape([2, -1]) :
 [[ 0  1  4]
 [ 9 16 25]]
x.reshape([3, -1]) :
 [[ 0  1]
 [ 4  9]
 [16 25]]


## Task

Reshape last vector to the shape `[3, 3]`

In [13]:
answer['reshaped'] = [
    ## YOUR CODE HERE
]

# 6. Slicing ND Arrays

We have recieved 2D array. You can work in numpy with 1D (vectors), 2D (matrices) and N-dimensional (tensors) arrays in the same way and using the same functions as for 1D arrays with one small exception: you can slice them with several slices.

For examplem for array

```
y = x.reshape([2, 3])
```

You can do the following slicing:

```
y[:, 1:]
```
That will select all the elements from all the rows (1st `:` indexing) and all the elements from the columns starting from 1st colom (second `1:` indexing):

In [14]:
y = x.reshape([2, 3])
print('y : \n', y)
print('y[:, 1:] :  \n', y[:, 1:])

y : 
 [[ 0  1  4]
 [ 9 16 25]]
y[:, 1:] :  
 [[ 1  4]
 [16 25]]


## Task

Reshape your last vector to $3 \times 3$ matrix and select 
- first two columns
- firts two rows
- first and last rows
- first and last columns

In [15]:

answer['sliced_reshaped'] = [
    ## YOUR CODE HERE
]   

# 6. Changing array dtype

You can also change dtype of the arrays, for instance `x.astype('int8')` will convert `x` to `int8`

In [16]:
print('x : ', x)
print("x.astype('int8') : ", x.astype('int8'))
print("x.astype('int8') : ", x.astype('bool'))

x :  [ 0  1  4  9 16 25]
x.astype('int8') :  [ 0  1  4  9 16 25]
x.astype('int8') :  [False  True  True  True  True  True]


## Task

Turn all of your vectors to boolean and then to float32

In [17]:
answer['bool'] = [
    ## YOUR CODE HERE
]

# 7. Very useful visualization

There is a very conveniet lib that helps reading large arrays, it is called `lovely_numpy` (and there is `lovely_tensors` for torch tensors)

You can print all the important info about your tensor using it:

In [18]:
import lovely_numpy as ln
print('ln.lovely(x) : ', ln.lovely(x))

ln.lovely(x) :  array[6] i64 x∈[0, 25] μ=9.167 σ=8.896 [0, 1, 4, 9, 16, 25]


Here 
- `array[6]` means the array that contains 6 elements
- `i64` means int64 type
- `x∈[0, 25]` is range of values
- `μ=9.167 σ=8.896` is mean and standard deviation for this array

## Task

Create `lovely_numpy` descriptions for the tensors in answer['vectors']

In [19]:
answer['lovely'] = [
    ## YOUR CODE HERE
]

# 8. Operations with arrays

You can perform binary elementwise operations with tensors as long as the tensors match in shape

For instance, there are the following binary operations:
- `+` -- sum
- `-` -- difference
- `*` -- product
- `/` -- fraction
- `%` -- calculate remainder
- `//` -- whole part of division
- `**` -- power
- `>` -- comparison

And there is whole set of the corresponding inplace operations

Actually, you may think of using any python binary operators for numpy arrays

In [ ]:
(x ** 2 + x ** 3) % 3

[           inf 3.00000000e+00 5.00000000e-01 8.13443073e-01
 1.60625000e+01 2.14752365e+03]


/var/folders/wl/b3bt1t9j3xg0_d35knd6stzr0000gn/T/ipykernel_26836/1229559742.py:2: RuntimeWarning: divide by zero encountered in divide
  print((x**2 +  2**x)/x**3)


## Task

Calculate the following expression for every vector in `answer['vetors']`:

$$(x^2 + 2^x) / x^3$$

In [35]:
answer['expression'] = [
    (x**2 + 2**x)/x**3
]

/var/folders/wl/b3bt1t9j3xg0_d35knd6stzr0000gn/T/ipykernel_26836/264812984.py:2: RuntimeWarning: divide by zero encountered in divide
  (x**2 + 2**x)/x**3


# 9. Pointwise operations with arrays

Lastly, you can perform any elementwise operations with arrays, such as:
- $\sin(x)$
- $\cos(x)$
- $\sqrt{x}$
- and many others

In [32]:
print(np.sin(x))
print(np.cos(x))
print(np.sqrt(x))

[ 0.          0.84147098 -0.7568025   0.41211849 -0.28790332 -0.13235175]
[ 1.          0.54030231 -0.65364362 -0.91113026 -0.95765948  0.99120281]
[0. 1. 2. 3. 4. 5.]


## Task

Calculate the following expression for every of the tensors in `answer['vectors']`:
$\sqrt{\sin^2(x) + \cos^2(x)}$

In [33]:
answer['expression'] = [
    np.sqrt(np.power(np.sin(x),2) + np.power(np.cos(x),2))
]

# 10. Save the answer

In [36]:
json_tricks.dump(answer, '.answer.json')

ValueError: Out of range float values are not JSON compliant: inf